internal/util/ssa/filterintent.go

```diff
@@ -0,0 +1,147 @@
+	// filter out changes not in the allowed paths (fields to not consider, e.g. status);
+	if len(input.AllowedPaths) > 0 {
+		FilterIntent(&FilterIntentInput{
+			Path:         contract.Path{},
+			Value:        obj.Object,
+			ShouldFilter: IsNotAllowedPath(input.AllowedPaths),
+		})
+	}
+
+	// filter out changes for ignore paths (well known fields owned by other controllers, e.g.
+	//   spec.controlPlaneEndpoint in the InfrastructureCluster object);
+	if len(input.IgnorePaths) > 0 {
+		FilterIntent(&FilterIntentInput{
+			Path:         contract.Path{},
+			Value:        obj.Object,
+			ShouldFilter: IsIgnorePath(input.IgnorePaths),
+		})
+	}
+}
+
+// FilterIntent ensures that object only includes the fields and values for which the controller has an opinion,
+// and filter out everything else by removing it from the Value.
+// NOTE: This func is called recursively only for fields of type Map, but this is ok given the current use cases
+// this func has to address. More specifically, we are using this func for filtering out not allowed paths and for ignore paths;
+// all of them are defined in reconcile_state.go and are targeting well-known fields inside nested maps.
+// Allowed paths / ignore paths which point to an array are not supported by the current implementation.
+func FilterIntent(ctx *FilterIntentInput) bool {
+	value, ok := ctx.Value.(map[string]interface{})
+	if !ok {
+		return false
+	}
+
+	gotDeletions := false
+	for field := range value {
+		fieldCtx := &FilterIntentInput{
+			// Compose the Path for the nested field.
+			Path: ctx.Path.Append(field),
+			// Gets the original and the modified Value for the field.
+			Value: value[field],
+			// Carry over global values from the context.
+			ShouldFilter: ctx.ShouldFilter,
+		}
+
+		// If the field should be filtered out, delete it from the modified object.
+		if fieldCtx.ShouldFilter(fieldCtx.Path) {
+			delete(value, field)
+			gotDeletions = true
+			continue
+		}
+
+		// Process nested fields and get in return if FilterIntent removed fields.
+		if FilterIntent(fieldCtx) {
+			// Ensure we are not leaving empty maps around.
+			if v, ok := fieldCtx.Value.(map[string]interface{}); ok && len(v) == 0 {
+				delete(value, field)
+				gotDeletions = true
+			}
+		}
+	}
+	return gotDeletions
+}
+
+// FilterIntentInput holds info required while filtering the intent for server side apply.
+// NOTE: in server side apply an intent is a partial object that only includes the fields and values for which the user has an opinion.
+type FilterIntentInput struct {
+	// the Path of the field being processed.
+	Path contract.Path
+
+	// the Value for the current Path.
+	Value interface{}
+
+	// ShouldFilter handle the func that determine if the current Path should be dropped or not.
+	ShouldFilter func(path contract.Path) bool
+}
+
+// IsAllowedPath returns true when the Path is one of the AllowedPaths.
+func IsAllowedPath(allowedPaths []contract.Path) func(path contract.Path) bool {
```

Comment:
```suggestion
// IsPathAllowed returns true when the Path is one of the AllowedPaths.
func IsPathAllowed(allowedPaths []contract.Path) func(path contract.Path) bool {
```

Reply:
Since these functions where not introduced in this package and just migrated from another internal package I would like to move this discussion into a follow-up issue to unblock this PR.

Since these functions are internal we are safe to change them in a follow-up.

I added a follow-up item for this in the umbrella issue.

(Same for the other ssa util review comments).

Reply:
+1

Reply:
+1 to fix this/similar methods in a follow-up now that those methods are part of our internal utils